### PR TITLE
CodeCamp #46

### PR DESCRIPTION
**Pull request by shufanwu**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

Motivation
models part-10 and others of [#1982](https://github.com/open-mmlab/mmdetection3d/discussions/1982)

Modification
Add typehint for models/utils, utils/array_converter.py and models/test_time_augs/merge_augs.py

BC-breaking (Optional)
Does the modification introduce changes that break the back-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

Use cases (Optional)
If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

Checklist
Pre-commit or other linting tools are used to fix the potential lint issues.
The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
If the modification has potential influence on downstream projects, this PR should be tested with downstream projects.
The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmdet3d/models/utils/clip_sigmoid.py [hunk 1]
```diff
@@ -1,17 +1,20 @@
...
 
-def clip_sigmoid(x, eps=1e-4):
+def clip_sigmoid(x: Tensor, eps: Optional[float] = 1e-4) -> Tensor:
     """Sigmoid function for input feature.
 
     Args:
-        x (torch.Tensor): Input feature map with the shape of [B, N, H, W].
-        eps (float, optional): Lower bound of the range to be clamped to.
+        x (Tensor): Input feature map with the shape of [B, N, H, W].
+        eps (float): Lower bound of the range to be clamped to.
```
Comment (JingweiZhang12) [inline mmdet3d/models/utils/clip_sigmoid.py:13]:
```suggestion
        eps (float, optional): Lower bound of the range to be clamped to.
```

File: mmdet3d/models/utils/edge_indices.py [hunk 2]
```diff
@@ -1,35 +1,38 @@
...
             image size, scaling factor, etc.
         downsample_ratio (int): Downsample ratio of output feature,
-        step (int, optional): Step size used for generateing
+        step (int): Step size used for generateing
             edge indices. Default: 1.
-        pad_mode (str, optional): Padding mode during data pipeline.
+        pad_mode (str): Padding mode during data pipeline.
             Default: 'default'.
-        dtype (torch.dtype, optional): Dtype of edge indices tensor.
+        dtype (torch.dtype): Dtype of edge indices tensor.
```
Comment (JingweiZhang12) [inline mmdet3d/models/utils/edge_indices.py:29]:
```suggestion
        dtype (np.dtype): Dtype of edge indices tensor.
```

File: mmdet3d/models/utils/gaussian.py [hunk 3]
```diff
@@ -22,17 +25,20 @@ def gaussian_2d(shape, sigma=1):
...
     """Get gaussian masked heatmap.
 
     Args:
-        heatmap (torch.Tensor): Heatmap to be masked.
-        center (torch.Tensor): Center coord of the heatmap.
+        heatmap (Tensor): Heatmap to be masked.
+        center (Tensor): Center coord of the heatmap.
         radius (int): Radius of gaussian.
-        K (int, optional): Multiple of masked_gaussian. Defaults to 1.
+        k (int): Multiple of masked_gaussian. Defaults to 1.
```
Comment (JingweiZhang12) [inline mmdet3d/models/utils/gaussian.py:38]:
```suggestion
        k (int, optional): Multiple of masked_gaussian. Defaults to 1.
```

File: mmdet3d/models/utils/gaussian.py [hunk 4]
```diff
@@ -54,15 +60,16 @@ def draw_heatmap_gaussian(heatmap, center, radius, k=1):
...
-def gaussian_radius(det_size, min_overlap=0.5):
+def gaussian_radius(det_size: Tuple[Tensor, Tensor],
+                    min_overlap: float = 0.5) -> Tensor:
     """Get radius of gaussian.
 
     Args:
-        det_size (tuple[torch.Tensor]): Size of the detection result.
-        min_overlap (float, optional): Gaussian_overlap. Defaults to 0.5.
+        det_size (Tuple[Tensor]): Size of the detection result.
+        min_overlap (float): Gaussian_overlap. Defaults to 0.5.
```
Comment (JingweiZhang12) [inline mmdet3d/models/utils/gaussian.py:69]:
```suggestion
        min_overlap (float, optional): Gaussian_overlap. Defaults to 0.5.
```

File: mmdet3d/models/utils/gaussian.py [hunk 5]
```diff
@@ -86,24 +93,28 @@ def gaussian_radius(det_size, min_overlap=0.5):
...
 
     Args:
         heatmap (Tensor): Input heatmap, the gaussian kernel will cover on
             it and maintain the max value.
-        center (list[int]): Coord of gaussian kernel's center.
+        center (List[int]): Coord of gaussian kernel's center.
         radius_x (int): X-axis radius of gaussian kernel.
         radius_y (int): Y-axis radius of gaussian kernel.
-        k (int, optional): Coefficient of gaussian kernel. Default: 1.
+        k (int): Coefficient of gaussian kernel. Default: 1.
```
Comment (JingweiZhang12) [inline mmdet3d/models/utils/gaussian.py:109]:
```suggestion
        k (int, optional): Coefficient of gaussian kernel. Default: 1.
```
These arguments which have default values are optional. Please check all the files.

File: mmdet3d/utils/array_converter.py [hunk 6]
```diff
@@ -250,26 +261,32 @@ def set_template(self, array):
...
 
         Raises:
             ValueError: If input is list or tuple and cannot be converted to
                 to a NumPy array, a ValueError is raised.
             TypeError: If input type does not belong to the above range,
                 or the contents of a list or tuple do not share the
                 same data type, a TypeError is raised.
+
+        Returns:
+            converted_array (np.ndarray | torch.Tensor): Array which
```
Comment (JingweiZhang12) [inline mmdet3d/utils/array_converter.py:288]:
```suggestion
            converted_array (np.ndarray | torch.Tensor): The converted array.
```

File: mmdet3d/utils/array_converter.py [hunk 7]
```diff
@@ -200,16 +204,23 @@ def recursive_recover(input_data):
...
 class ArrayConverter:
-
     SUPPORTED_NON_ARRAY_TYPES = (int, float, np.int8, np.int16, np.int32,
                                  np.int64, np.uint8, np.uint16, np.uint32,
                                  np.uint64, np.float16, np.float32, np.float64)
 
-    def __init__(self, template_array=None):
+    def __init__(self,
+                 template_array: Optional[TemplateArrayType] = None) -> None:
+        """Utility class for data-type agnostic processing.
```
Comment (JingweiZhang12) [inline mmdet3d/utils/array_converter.py:213]: Please add docstring above this '__init__' func

File: mmdet3d/models/test_time_augs/merge_augs.py [hunk 8]
```diff
@@ -1,22 +1,27 @@
 # Copyright (c) OpenMMLab. All rights reserved.
+from typing import List
+
 import torch
 
 from mmdet3d.structures import bbox3d2result, bbox3d_mapping_back, xywhr2xyxyr
+from ...utils import OptConfigType
```
Comment (JingweiZhang12) [inline mmdet3d/models/test_time_augs/merge_augs.py:7]:
```suggestion
from mmdet3d.utils import OptConfigType
```

**Comment by JingweiZhang12** [comment]
@Xiangxu-0103, please review this PR

File: mmdet3d/models/test_time_augs/merge_augs.py [hunk 9]
```diff
@@ -1,22 +1,27 @@
...
 
 from mmdet3d.structures import bbox3d2result, bbox3d_mapping_back, xywhr2xyxyr
+from mmdet3d.utils import OptConfigType
 from ..layers import nms_bev, nms_normal_bev
 
 
-def merge_aug_bboxes_3d(aug_results, aug_batch_input_metas, test_cfg):
+def merge_aug_bboxes_3d(aug_results: List[dict],
+                        aug_batch_input_metas: List[dict],
+                        test_cfg: OptConfigType) -> dict:
```
Comment (Xiangxu-0103) [inline mmdet3d/models/test_time_augs/merge_augs.py:13]:
```suggestion
                        test_cfg: ConfigType) -> dict:
```

File: mmdet3d/models/test_time_augs/merge_augs.py [hunk 10]
```diff
@@ -1,22 +1,27 @@
...
 
             - bbox_3d (:obj:`BaseInstance3DBoxes`): Detection bbox.
-            - scores_3d (torch.Tensor): Detection scores.
-            - labels_3d (torch.Tensor): Predicted box labels.
-        img_metas (list[dict]): Meta information of each sample.
-        test_cfg (dict): Test config.
+            - scores_3d (Tensor): Detection scores.
+            - labels_3d (Tensor): Predicted box labels.
+        aug_batch_input_metas (List[dict]): Meta information of each sample.
+        test_cfg (:obj:`OptConfigType`): Test config.
```
Comment (Xiangxu-0103) [inline mmdet3d/models/test_time_augs/merge_augs.py:24]:
```suggestion
        test_cfg (dict or :obj:`ConfigDict`): Test config.
```

File: mmdet3d/models/utils/clip_sigmoid.py [hunk 11]
```diff
@@ -1,17 +1,20 @@
 # Copyright (c) OpenMMLab. All rights reserved.
+from typing import Optional
+
 import torch
+from torch import Tensor
 
 
-def clip_sigmoid(x, eps=1e-4):
+def clip_sigmoid(x: Tensor, eps: Optional[float] = 1e-4) -> Tensor:
```
Comment (Xiangxu-0103) [inline mmdet3d/models/utils/clip_sigmoid.py:8]:
```suggestion
def clip_sigmoid(x: Tensor, eps: float = 1e-4) -> Tensor:
```

File: mmdet3d/models/utils/edge_indices.py [hunk 12]
```diff
@@ -1,35 +1,38 @@
...
-                     downsample_ratio,
-                     step=1,
-                     pad_mode='default',
-                     dtype=np.float32,
-                     device='cpu'):
+def get_edge_indices(img_metas: List[dict],
+                     downsample_ratio: int,
+                     step: int = 1,
+                     pad_mode: str = 'default',
+                     dtype: np.dtype = np.float32,
```
Comment (Xiangxu-0103) [inline mmdet3d/models/utils/edge_indices.py:13]:
```suggestion
                     dtype: type = np.float32,
```

File: mmdet3d/models/utils/edge_indices.py [hunk 13]
```diff
@@ -1,35 +1,38 @@
...
     with CUDA tensor (0.09s and 0.72s in 100 runs).
 
     Args:
-        img_metas (list[dict]): Meta information of each image, e.g.,
+        img_metas (List[dict]): Meta information of each image, e.g.,
             image size, scaling factor, etc.
         downsample_ratio (int): Downsample ratio of output feature,
-        step (int, optional): Step size used for generateing
+        step (int): Step size used for generateing
             edge indices. Default: 1.
```
Comment (Xiangxu-0103) [inline mmdet3d/models/utils/edge_indices.py:26]:
```suggestion
            edge indices. Defaults to 1.
```

File: mmdet3d/models/utils/edge_indices.py [hunk 14]
```diff
@@ -1,35 +1,38 @@
...
-        pad_mode (str, optional): Padding mode during data pipeline.
+        pad_mode (str): Padding mode during data pipeline.
             Default: 'default'.
```
Comment (Xiangxu-0103) [inline mmdet3d/models/utils/edge_indices.py:28]:
```suggestion
            Defaults to 'default'.
```

File: mmdet3d/models/utils/gaussian.py (hunk 5 above)
Comment (Xiangxu-0103) [reply mmdet3d/models/utils/gaussian.py:109]: It seems that only default value is `None` should be optional in the latest version.

File: mmdet3d/utils/array_converter.py [hunk 15]
```diff
@@ -200,16 +204,23 @@ def recursive_recover(input_data):
...
-
     SUPPORTED_NON_ARRAY_TYPES = (int, float, np.int8, np.int16, np.int32,
                                  np.int64, np.uint8, np.uint16, np.uint32,
                                  np.uint64, np.float16, np.float32, np.float64)
+    """Utility class for data-type agnostic processing.
+
+    Args:
+        template_array (tuple | list | int | float | np.ndarray |
+        torch.Tensor, optional):
+            template array.
```
Comment (Xiangxu-0103) [inline mmdet3d/utils/array_converter.py:215]:
```suggestion
            template array. Defaults to None.
```

File: mmdet3d/utils/array_converter.py [hunk 16]
```diff
@@ -250,26 +261,31 @@ def set_template(self, array):
             raise TypeError(f'Template type {self.array_type}'
                             f' is not supported.')
 
-    def convert(self, input_array, target_type=None, target_array=None):
+    def convert(
+            self,
+            input_array: TemplateArrayType,
+            target_type: TargetType = None,
+            target_array: TargetType = None
```
Comment (Xiangxu-0103) [inline mmdet3d/utils/array_converter.py:268]:
```suggestion
            target_array: OptArrayType = None
```

File: mmdet3d/utils/array_converter.py [hunk 17]
```diff
@@ -250,26 +261,31 @@ def set_template(self, array):
...
-            target_type (<class 'np.ndarray'> | <class 'torch.Tensor'>,
-                optional):
-                Type to which input array is converted. Defaults to None.
+            input_array (tuple | list | int | float | np.ndarray |
+            torch.Tensor): Input array.
+            target_type (np.ndarray | torch.Tensor, optional):
+                Type to which input array is converted.
             target_array (np.ndarray | torch.Tensor, optional):
                 Template array to which input array is converted.
-                Defaults to None.
```
Comment (Xiangxu-0103) [inline mmdet3d/utils/array_converter.py:265]: keep this line.

File: mmdet3d/utils/array_converter.py [hunk 18]
```diff
@@ -250,26 +261,31 @@ def set_template(self, array):
...
-            input_array (tuple | list | np.ndarray |
-                torch.Tensor | int | float ):
-                Input array. Defaults to None.
-            target_type (<class 'np.ndarray'> | <class 'torch.Tensor'>,
-                optional):
-                Type to which input array is converted. Defaults to None.
+            input_array (tuple | list | int | float | np.ndarray |
+            torch.Tensor): Input array.
+            target_type (np.ndarray | torch.Tensor, optional):
+                Type to which input array is converted.
```
Comment (Xiangxu-0103) [inline mmdet3d/utils/array_converter.py:276]:
```suggestion
                Type to which input array is converted. Defaults to None.
```

File: mmdet3d/utils/array_converter.py [hunk 19]
```diff
@@ -250,26 +261,31 @@ def set_template(self, array):
...
 
         Raises:
             ValueError: If input is list or tuple and cannot be converted to
                 to a NumPy array, a ValueError is raised.
             TypeError: If input type does not belong to the above range,
                 or the contents of a list or tuple do not share the
                 same data type, a TypeError is raised.
+
+        Returns:
+            converted_array (np.ndarray | torch.Tensor): The converted array.
```
Comment (Xiangxu-0103) [inline mmdet3d/utils/array_converter.py:288]:
```suggestion
            np.ndarray or torch.Tensor: The converted array.
```

File: mmdet3d/utils/array_converter.py [hunk 20]
```diff
@@ -309,7 +325,17 @@ def convert(self, input_array, target_type=None, target_array=None):
...
+    def recover(
+        self, input_array: Union[np.ndarray, torch.Tensor]
+    ) -> Union[np.ndarray, torch.Tensor]:
+        """Recover input type to original array type.
+
+        Args:
+            input_array (np.ndarray | torch.Tensor): Input array.
+
+        Returns:
+            converted_array (np.ndarray | torch.Tensor): Converted array.
```
Comment (Xiangxu-0103) [inline mmdet3d/utils/array_converter.py:337]:
```suggestion
            np.ndarray or torch.Tensor: Converted array.
```

File: mmdet3d/utils/array_converter.py [hunk 21]
```diff
@@ -250,15 +260,19 @@ def set_template(self, array):
...
-            input_array (tuple | list | np.ndarray |
-                torch.Tensor | int | float ):
-                Input array. Defaults to None.
-            target_type (<class 'np.ndarray'> | <class 'torch.Tensor'>,
-                optional):
+            input_array (tuple | list | int | float | np.ndarray |
+                torch.Tensor): Input array.
+            target_type (:class:`np.ndarray` or :class:`torch.Tensor`,
+                optional): Type to which input array is converted.
                 Type to which input array is converted. Defaults to None.
```
Comment (JingweiZhang12) [inline mmdet3d/utils/array_converter.py:276]: Repeated docstrings.

**Comment by JingweiZhang12** [comment]
Please fix the lint.

**Review by JingweiZhang12 — APPROVED** [review]

**Review by Xiangxu-0103 — APPROVED** [review]

**Pull request merged**